lib/wallet/src/Cardano/Wallet/DB/Store/Transactions/Model.hs

```diff
@@ -284,59 +255,99 @@ mkTxRelation tx =
+    fromTxOutToken token =
+        ( AssetId (txOutTokenPolicyId token) (txOutTokenName token)
+        , txOutTokenQuantity token
+        )
+
+fromTxCollateralOut :: (TxCollateralOut, [TxCollateralOutToken]) -> W.TxOut
+fromTxCollateralOut (out,tokens) =
+    W.TxOut
+    { W.address = txCollateralOutAddress out
+    , W.tokens = TokenBundle.fromFlatList
+            (txCollateralOutAmount out)
+            (fromTxCollateralOutToken <$> tokens)
+    }
+  where
+    fromTxCollateralOutToken token =
+        ( AssetId
+            (txCollateralOutTokenPolicyId token)
+            (txCollateralOutTokenName token)
+        , txCollateralOutTokenQuantity token
+        )
+
+{-------------------------------------------------------------------------------
+    Decorating Tx inputs with outputs
+-------------------------------------------------------------------------------}
 type TxOutKey = (TxId, Word32)
 
-decorateWithTxOuts :: TxHistoryF 'Without -> TxHistoryF 'With
-decorateWithTxOuts (TxHistoryF w) =
-    let
-        txouts :: Map TxOutKey (TxOut, [TxOutToken])
-        txouts = Map.fromList $ do
-            TxRelationF {..} <- toList w
-            [(txOutputTxId &&& txOutputIndex $ txout, x) | x@(txout,_ ) <- outs]
-    in  TxHistoryF $ fmap (solveTxOut txouts) w
-
-decorateInputs
-    :: (t -> TxOutKey)
-    -> Map TxOutKey (TxOut, [TxOutToken])
-    -> [t]
-    -> [WithTxOut t]
-decorateInputs keyOf txOutMap ins = do
-        i <- ins
-        pure $ WithTxOut i $ Map.lookup (keyOf i) txOutMap
-
-solveTxOut
-    :: Map TxOutKey (TxOut, [TxOutToken])
-    -> TxRelationF 'Without
-    -> TxRelationF 'With
-solveTxOut txOutMap TxRelationF {..} = TxRelationF
-    { ins =
-        decorateInputs
-            (txInputSourceTxId &&& txInputSourceIndex)
-            txOutMap
-            ins
-    , collateralIns =
-        decorateInputs
-            (txCollateralSourceTxId &&& txCollateralSourceIndex)
-            txOutMap
-            collateralIns
-    , outs = outs
-    , collateralOuts = collateralOuts
-    , withdrawals = withdrawals
-    }
+toKeyTxIn :: TxIn -> TxOutKey
+toKeyTxIn txin = (txInputSourceTxId txin, txInputSourceIndex txin)
 
-undecorateFromTxOuts :: TxHistoryF 'With -> TxHistoryF 'Without
-undecorateFromTxOuts (TxHistoryF w) = TxHistoryF $ fmap unsolveTxOut w
+toKeyTxCollateral :: TxCollateral -> TxOutKey
+toKeyTxCollateral txcol =
+    (txCollateralSourceTxId txcol, txCollateralSourceIndex txcol)
 
-unsolveTxOut :: TxRelationF 'With -> TxRelationF 'Without
-unsolveTxOut TxRelationF {..} = TxRelationF
-    { ins = fmap txIn ins
-    , collateralIns = fmap txIn collateralIns
-    , outs = outs
-    , collateralOuts = collateralOuts
-    , withdrawals = withdrawals
+-- | A collection of Tx inputs
+-- (regular or collateral, refered to by input and order)
+-- that are decorated with the values of their corresponding Tx outputs.
+newtype DecoratedTxIns = DecoratedTxIns
+    { unDecoratedTxIns
+        :: Map TxOutKey W.TxOut
```

Comment:
Not changed in the PR, but why `TxOutKey` and not just `TxIn`?

Reply:
I agree — please don't ask. 😂 I'd like to move `TxIn` into the `Read` hierarchy first.

As `DecoratedTxIns` is exported as an abstract data type, I believe it is alright to let it use any wacky type internally at the moment — the `TxOutKey` is not part of its interface. 😏